flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/GenericInMemoryCatalog.java

```diff
@@ -209,11 +209,17 @@ public void alterTable(ObjectPath tablePath, CatalogBaseTable newTable, boolean
 		checkNotNull(tablePath);
 		checkNotNull(newTable);
 
-		// TODO: validate the new and old CatalogBaseTable must be of the same type. For example, this doesn't
-		//		allow alter a regular table to partitioned table, or alter a view to a table, and vice versa.
-		//		And also add unit tests.
-
 		if (tableExists(tablePath)) {
+			CatalogBaseTable oldTable = tables.get(tablePath);
+
+			// Theoretically, in-memory catalog can store any tables and views implementations
+			// thus, we need to check they are of the same class
+			if (oldTable.getClass() != newTable.getClass()) {
+				throw new CatalogException(
+					String.format("Existing table is '%s' and new table is '%s'. They should be of the same class.",
```

Comment:
Same as above. We can probably say "Table types doesn't match. ..."

Reply:
see above